### PR TITLE
HDDS-6312. Use KeyPrefixContainer table to accelerate the process of DELETE/UPDATE events

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

Recon stores the mapping of ContainerKeyPrefix in local RocksDB. When Recon is applying DELETE or UPDATE events from OM, it will run search the whole table for each to_be_deleted record.

In a big cluster, the record count in this table could be very large, and the search loop for each records is very slow. In our cluster there are 90m records, each loop cost over 70 seconds, if a delta OM events have 100 DELETE or UPDATE events, it will took about two hours to apply these updates.

This ticket is to accelerate the process with the help of a new local table.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6312

## How was this patch tested?

unit test.

### CODE REVIEW

**Comment by symious** [comment]
@adoroszlai @avijayanhwx @ferhui Could you help to review this PR?

**Comment by ferhui** [comment]
@JacksonYao287 Could you please help review?

**Comment by symious** [comment]
@avijayanhwx Could you help to check this issue?
This issue can be a problem if we want to retrieve updated information from OM.

In a small cluster, the connection from Recon to OM can be quite normal, with 10 minutes' interval. Please ignore the content in the following block, but each request is from Recon to OM to get delta updates.
```
2022-04-06 09:41:57,273 [Socket Reader #1 for port 9862] WARN org.apache.hadoop.ipc.Server: Connection Authentication from Recon:56662 for protocol org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol is failed for user ozone (auth:SIMPLE). 
2022-04-06 09:51:57,357 [Socket Reader #1 for port 9862] WARN org.apache.hadoop.ipc.Server: Connection Authentication from Recon:60932 for protocol org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol is failed for user ozone (auth:SIMPLE). 
2022-04-06 10:01:57,436 [Socket Reader #1 for port 9862] WARN org.apache.hadoop.ipc.Server: Connection Authentication from Recon:63884 for protocol org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol is failed for user ozone (auth:SIMPLE). 
2022-04-06 10:11:57,544 [Socket Reader #1 for port 9862] WARN org.apache.hadoop.ipc.Server: Connection Authentication from Recon:24028 for protocol org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol is failed for user ozone (auth:SIMPLE). 
2022-04-06 10:22:04,916 [Socket Reader #1 for port 9862] WARN org.apache.hadoop.ipc.Server: Connection Authentication from Recon:26968 for protocol org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol is failed for user ozone (auth:SIMPLE). 
```
But in a big cluster, since the process is too slow, this connection can be far longer then 10min's interval.

**Comment by adoroszlai** [comment]
Thanks @symious for the patch and sorry for the long delay in review.  Can you please resolve merge conflicts?

**Comment by symious** [comment]
@adoroszlai Thanks for the review. Updated the patch, please have a check.

**Comment by smengcl** [comment]
Thanks @symious for the patch. There's another patch causing minor conflicts. Would you merge latest master to the PR branch again?

I have resolved the conflict on my branch (I tried to push to your PR branch but got no permission), you can take this as a reference: https://github.com/smengcl/hadoop-ozone/blob/HDDS-6312/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java#L249-L275

**Comment by symious** [comment]
@smengcl Updated the PR, please have a look.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 1]
```diff
@@ -252,48 +253,54 @@ private void handleDeleteOMKeyEvent(String key,
...
         if (keyPrefix.equals(key)) {
-          keysToBeDeleted.add(keyValue.getKey());
+          if (keyValue.getKey().getContainerId() != -1) {
+            keysToBeDeleted.add(keyValue.getKey().toContainerKeyPrefix());
+          }
+        } else {
+          break;
         }
       }
-    }
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:269]: Hmm, why not release `keyContainerIterator` here?
Do we have other reasons to keep `keyContainerIterator` til the method end?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java [hunk 2]
```diff
@@ -97,13 +102,19 @@ public void reinitWithNewContainerDataFromOm(Map<ContainerKeyPrefix, Integer>
       throws IOException {
     // clear and re-init all container-related tables
     truncateTable(this.containerKeyTable);
+    truncateTable(this.keyContainerTable);
     truncateTable(this.containerKeyCountTable);
     initializeTables();
 
     if (containerKeyPrefixCounts != null) {
+      KeyPrefixContainer tmpKeyPreifxContainer;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java:110]: nit: typo
```suggestion
      KeyPrefixContainer tmpKeyPrefixContainer;
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java [hunk 3]
```diff
@@ -515,4 +547,84 @@ public void commitBatchOperation(RDBBatchOperation rdbBatchOperation)
...
+   * @param keyVersion the given keyVersion.
+   * @return Map of (KeyPrefixContainer, Integer).
+   */
+  @Override
+  public Map<KeyPrefixContainer, Integer> getContainerForKeyPrefixes(
+      String keyPrefix, long keyVersion) throws IOException {
+
+    Map<KeyPrefixContainer, Integer> containers = new LinkedHashMap<>();
+    TableIterator<KeyPrefixContainer, ? extends KeyValue<KeyPrefixContainer,
+        Integer>> keyIterator = keyContainerTable.iterator();
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java:565]: Need to close DB table iterator after use.

Either do `keyIterator.close()` before each `return`, or use try-with-resource (recommended).

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java [hunk 4]
```diff
@@ -515,4 +547,84 @@ public void commitBatchOperation(RDBBatchOperation rdbBatchOperation)
...
+          containers.put(new KeyPrefixContainer(keyPrefix,
+                  keyPrefixContainer.getKeyVersion(),
+                  keyPrefixContainer.getContainerId()),
+              keyValue.getValue());
+        } else {
+          LOG.warn("Null container returned for keyPrefix = {}," +
+                  " keyVersion = {} ", keyPrefix, keyVersion);
+        }
+      } else {
+        break; //Break when the first mismatch occurs.
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java:605]: nit: comment style
```suggestion
        // Break on first mismatch
        break;
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    KeyPrefixContainer that = (KeyPrefixContainer) o;
+    return (this.containerId == that.containerId) &&
+        this.keyPrefix.equals(that.keyPrefix) &&
+        this.keyVersion == that.keyVersion;
+  }
+
+  @Override
+  public int hashCode() {
+    return Long.valueOf(containerId).hashCode() + 13 * keyPrefix.hashCode() +
+        17 * Long.valueOf(keyVersion).hashCode();
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java:97]: I wondered why existing `ContainerKeyPrefix` did this.

Does this have any advantage over:
```java
return Objects.hash(containerId, keyPrefix, keyVersion);
```
?

**Review by smengcl — COMMENTED** [review]
Thanks @symious for the improvement.

Latest revision looks good overall. Minor comments inline.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/KeyPrefixContainerCodec.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @Override
+  public KeyPrefixContainer fromPersistedFormat(byte[] rawData)
+      throws IOException {
+
+    // When reading from byte[], we can always expect to have the key, version
+    // and version parts in the byte array.
+    byte[] keyBytes = ArrayUtils.subarray(rawData,
+        0, rawData.length - Long.BYTES * 2 - 2);
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/KeyPrefixContainerCodec.java:72]: Noted with thanks, I'll try to create a new ticket for this transition.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java (hunk 5 above)
Comment (symious) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java]: I think it's kind of coding habit?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java (hunk 3 above)
Comment (symious) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/ReconContainerMetadataManagerImpl.java]: Updated with try-with-resource.

**Comment by symious** [comment]
@smengcl Updated the patch, please have a look.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 1 above)
Comment (smengcl) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:269]: Hi @symious , other changes looks good now. Can we move the closing bracket (for the `try`) back here?

**Review by smengcl — APPROVED** [review]
+1 pending CI

**Comment by symious** [comment]
@smengcl Thank you for the review.

**Comment by symious** [comment]
@smengcl Do you have any suggestions on the failed unit test?

**Comment by smengcl** [comment]
> @smengcl Do you have any suggestions on the failed unit test?

The failure looks related at a first glance: https://github.com/apache/ozone/runs/8008158105

```bash
Error:  org.apache.hadoop.ozone.recon.recovery.TestReconOmMetadataManagerImpl.testUpdateOmDB  Time elapsed: 0.249 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at org.apache.hadoop.ozone.recon.recovery.TestReconOmMetadataManagerImpl.testUpdateOmDB(TestReconOmMetadataManagerImpl.java:139)
```

But [Line 139 in TestReconOmMetadataManagerImpl.java](https://github.com/symious/ozone/blob/HDDS-6312/hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/recovery/TestReconOmMetadataManagerImpl.java#L139) doesn't really match any meaningful code on your branch, hmm.

I ran the same test locally and it passed.

Retriggering the CI.

**Comment by szetszwo** [comment]
@symious , question -- what is the reason for adding `KeyPrefixContainer` but not reusing `ContainerKeyPrefix`?  I would like to see if we can get ride one of them.

**Comment by symious** [comment]
> In a big cluster, the record count in this table could be very large, and the search loop for each records is very slow. In our cluster there are 90m records, each loop cost over 70 seconds, if a delta OM events have 100 DELETE or UPDATE events, it will took about two hours to apply these updates."

@szetszwo It is to solve the slowness of processing.

**Comment by szetszwo** [comment]
I understand that this is a to fix the slowness.  Well done!

My question is why we cannot reuse `ContainerKeyPrefix`?  How `KeyPrefixContainer` can fix the slowness but  `ContainerKeyPrefix` cannot?

**Comment by symious** [comment]
> My question is why we cannot reuse ContainerKeyPrefix? How can KeyPrefixContainer fix the slowness but ContainerKeyPrefix cannot?

The slowness was coming from the operation of "getContainerForKeyPrefixes", since `ContainerKeyPrefix` is for mapping "container -> keyPrefix", so if we are trying to get the containers for a specific keyPreifx, we need to travel all the table of `ContainerKeyPrefix`, and it's time consuming when `ContainerKeyPrefix` is large.

With the help of "KeyPrefixContainer", the operation of "getContainerForKeyPrefixes" will be quite fast.

@szetszwo

**Comment by szetszwo** [comment]
@symious , sorry, I still cannot understand.  I probably have not asked the right question.  These two classes (shown below) look the same.  What we need to add `KeyPrefixContainer` but not reusing `ContainerKeyPrefix `?
```java
public class ContainerKeyPrefix {

  private long containerId;
  private String keyPrefix;
  private long keyVersion = -1;
```
```java
public class KeyPrefixContainer {

  private String keyPrefix;
  private long keyVersion = -1;
  private long containerId = -1;
```

> With the help of "KeyPrefixContainer", ...

How it helps?

**Comment by symious** [comment]
@szetszwo Oh, I got your question.

I think I created a new class for the readability.

It is kind of confusing to have 
```
  private Table<ContainerKeyPrefix, Integer> containerKeyTable;
  private Table<ContainerKeyPrefix, Integer> keyContainerTable;
```

**Comment by szetszwo** [comment]
> ...  for the readability.

If the code are the same, it is better to reuse the code.

```java
//SCMMetadataStoreImpl
  private Table<BigInteger, X509Certificate> validCertsTable;
  private Table<BigInteger, X509Certificate> validSCMCertsTable;
  private Table<BigInteger, X509Certificate> revokedCertsTable;
```
For examples, we won't create a different `BigInteger` and a different `X509Certificate` for the tables above.

**Comment by symious** [comment]
> If the code are the same, it is better to reuse the code.

Reasonable to me, also.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    KeyPrefixContainer that = (KeyPrefixContainer) o;
+    return (this.containerId == that.containerId) &&
+        this.keyPrefix.equals(that.keyPrefix) &&
+        this.keyVersion == that.keyVersion;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(containerId, keyPrefix, keyPrefix);
```
Comment (szetszwo) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java:98]: @symious , another question --  the repeated `keyPrefix, keyPrefix` looks like a typo.  For fixing it, we will replace the last `keyPrefix` with `keyVersion` as below.
```java
    return Objects.hash(containerId, keyPrefix, keyVersion);
```
However, the hash function will be changed.  Now the question is -- Is it an incompatible change?  Have the hash value used or written anywhere in the DB?

Comment (symious) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java:98]: Oh, thanks for spotting that.

I didin't find the persistence of the hash value. The only place I found being used should be in the "getContainerForKeyPrefixes" function, which should not cause incompatibilities.

Comment (szetszwo) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/KeyPrefixContainer.java:98]: @symious , thanks a lot for checking it!  Let me change it to use the same `hashCode()` as `ContainerKeyPrefix` in [HDDS-8733](https://issues.apache.org/jira/browse/HDDS-8733).

**Pull request merged**